iCubTelemVizServer/terminationHandler.js

```diff
@@ -45,23 +53,45 @@ TerminationHandler.prototype.runSubsetA = function (signal) {
                 reject(error);
             }
         });
-        console.log('Control Console Server closing: no further incoming requests accepted.');
+        console.log('Control Console Server closing: no further connection requests accepted.');
     }.bind(this));
-    closeChildProcessPromise.then(
-        function(value) {
-            console.log(value);
-            this.runSubsetB();
-        }.bind(this),
-        function(error) { console.error(error.toString()); }
-    );
+
+    // Return the promises
+    return {
+        closeChildProcess: closeChildProcessPromise,
+        closeServers: Promise.all([closeTelemServerPromise,closeConsoleServerPromise])
+    };
 }
 
-TerminationHandler.prototype.runSubsetB = function() {
+TerminationHandler.prototype.runSubsetB = function(resValue) {
     /*
      * === Closure subset B ===
      * B.1 - Stop listening to client requests ("subscribe"/"unsubscribe" events) on the existing connections.
+     * B.2 - Close the YARP ports connections associated to the telemetry entries in `portInConfig` and instantiated by the main Node.js process.
+     * B.3 - Stop reading the respective YARP ports (similar to turning off the port listeners).
+     * B.4 - Stop sending "real-time telemetry data" messages (notifications) to subscribers.
      */
-     this.telemServerTracker.pauseAll();
+     console.log(resValue);              // consume resolve value from previous promise (synch operation)
+     this.telemServerTracker.pauseAll(); // Stop listening to client requests (synch operation)
+     console.log('iCub Telemetry Server closing: no further "subscribe"/"unsubscribe" requests accepted.');
+     this.unlistenToYarpPorts.forEach((disconnect) => { disconnect(); }); // Disconnect all telemetry entries (asynch operation)
```

Comment:
Just FYI, when calling the ``close`` method on a port, the disconnections are handled automatically.

Reply:
Good point. I had the following reasons:
- Initially I was leaving the ports closure to the `Yarp.js` module which does it automatically when the process exits (https://github.com/robotology/yarp.js/blob/ceddda861b1b29478d3bc389d83765fed7297196/yarp.js#L201-L205).
- Nevertheless, I wanted the state of the Yarp ports to be consistent with the local disconnection on the Node.js side, and the calls to the Yarp port listeners to stop.

At this point, there is something left in the Node Event Loop, preventing the process to exit on its own (without explicitly calling `process.exit()`, which is not recommended), so I'm trying to see what happens if I close the ports manually (not committed here). I doubt but if that helps, I'll replace the ports disconnection by the direct closure.

Reply:
Ok clear, I was just curious.